File: apps/cli/src/commands/status.ts
import { type SessionMeta, Store } from "@teleprompter/daemon";
import { dim, green, red } from "../lib/colors";
import { isDaemonRunning } from "../lib/ensure-daemon";
import { formatAge } from "../lib/format";

/**
 * tp status — shows the current daemon state.
 *
 * Reads the Store directly (the Store is the source of truth regardless of
 * whether the background daemon is currently running). Also reports whether
 * the daemon IPC socket is live.
 */
export async function statusCommand(_argv: string[]): Promise<void> {
  const backgroundRunning = await isDaemonRunning();

  const store = new Store();
  const sessions = store.listSessions();
  store.close();

  displayStatus(sessions, backgroundRunning);
  process.exit(0);
}

function displayStatus(
  sessions: SessionMeta[],
  backgroundRunning: boolean,
): void {
  console.log("");
  console.log(`Daemon Status`);
  console.log(`─────────────`);
  console.log(
    `Background daemon: ${
      backgroundRunning ? green("running") : dim("not running")
    }`,
  );
  console.log(`Sessions: ${sessions.length}`);
  console.log("");

  if (sessions.length === 0) {
    console.log("No active sessions.");
    console.log("");
    console.log("Start a session:");
    console.log("  tp -p 'hello'                    # passthrough mode");
    console.log("  tp daemon start --spawn --cwd .   # managed mode");
    return;
  }

  const groups = new Map<string, SessionMeta[]>();
  for (const s of sessions) {
    const key = s.worktree_path ?? s.cwd;
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key)?.push(s);
  }

  for (const [path, group] of groups) {
    console.log(`  ${path}`);
    for (const s of group) {
      const indicator =
        s.state === "running"
          ? green("●")
          : s.state === "stopped"
            ? dim("○")
            : red("✕");

      console.log(`    ${indicator} ${s.sid}  seq=${s.last_seq}  ${s.state}`);
      if (s.claude_version) {
        console.log(`      claude ${s.claude_version}`);
      }
      const age = formatAge(Date.now() - s.updated_at);
      console.log(`      updated ${age}`);
    }
    console.log("");
  }
}

File: rust/tp-cli/src/commands/status.rs
//! `tp status` — daemon liveness + session list.
//!
//! Byte-exact port of `apps/cli/src/commands/status.ts`. Reads the Store
//! directly (source of truth regardless of whether the daemon is running) and
//! probes the IPC socket for liveness. Output: a "Daemon Status" block, the
//! background-daemon line (green "running" / dim "not running"), the session
//! count, then per-worktree groups with ●/○/✕ indicators and a "updated N ago"
//! line. Empty store prints "No active sessions." + start hints.

use std::process::ExitCode;

use crate::colors::{dim, green, red};
use crate::format::format_age;
use crate::socket::is_daemon_running;
use crate::store::{list_sessions, SessionRow};
use crate::util::now_ms;

pub fn run() -> ExitCode {
    let background_running = is_daemon_running();
    let sessions = list_sessions();
    display_status(&sessions, background_running);
    ExitCode::SUCCESS
}

fn display_status(sessions: &[SessionRow], background_running: bool) {
    println!();
    println!("Daemon Status");
    println!("─────────────");
    let daemon_state = if background_running {
        green("running")
    } else {
        dim("not running")
    };
    println!("Background daemon: {daemon_state}");
    println!("Sessions: {}", sessions.len());
    println!();

    if sessions.is_empty() {
        println!("No active sessions.");
        println!();
        println!("Start a session:");
        println!("  tp -p 'hello'                    # passthrough mode");
        println!("  tp daemon start --spawn --cwd .   # managed mode");
        return;
    }

    // Group by worktree_path ?? cwd, preserving first-seen order (Map insertion
    // order in the TS). We can't use a HashMap (loses order), so track keys in a
    // Vec alongside.
    let now = now_ms();
    let mut order: Vec<String> = Vec::new();
    let mut groups: std::collections::HashMap<String, Vec<&SessionRow>> =
        std::collections::HashMap::new();
    for s in sessions {
        let key = s.worktree_path.clone().unwrap_or_else(|| s.cwd.clone());
        if !groups.contains_key(&key) {
            order.push(key.clone());
        }
        groups.entry(key).or_default().push(s);
    }

    for path in &order {
        println!("  {path}");
        for s in &groups[path] {
            let indicator = match s.state.as_str() {
                "running" => green("●"),
                "stopped" => dim("○"),
                _ => red("✕"),
            };
            println!("    {indicator} {}  seq={}  {}", s.sid, s.last_seq, s.state);
            if let Some(ver) = &s.claude_version {
                println!("      claude {ver}");
            }
            let age = format_age(now - s.updated_at, now);
            println!("      updated {age}");
        }
        println!();
    }
}
